src/realm/views.cpp

```diff
@@ -190,9 +199,121 @@ bool SortDescriptor::Sorter::operator()(IndexPair i, IndexPair j, bool total_ord
     return total_ordering ? i.index_in_view < j.index_in_view : 0;
 }
 
-void RowIndexes::do_sort(const SortDescriptor& order, const SortDescriptor& distinct)
+void DescriptorOrdering::append_sort(SortDescriptor sort)
+{
+    size_t orig_length = m_descriptors.size();
+    if (orig_length == 0) {
+        m_descriptors.emplace_back(std::move(sort));
+    }
+    else if (descriptor_is_sort(orig_length - 1)) {
+        // merge them together by appending internally
+        m_descriptors[orig_length - 1].merge_with(std::move(sort));
+    }
+    else { // last descriptor is distinct, just add it
+        m_descriptors.emplace_back(std::move(sort));
+    }
+}
+
+void DescriptorOrdering::append_distinct(DistinctDescriptor distinct)
+{
+    size_t orig_length = m_descriptors.size();
+    if (orig_length == 0) {
+        // add an empty sort descriptor first to keep ordering
+        m_descriptors.emplace_back(SortDescriptor());
+        m_descriptors.emplace_back(std::move(distinct));
```

Comment:
Hmm good point. Don't you think it reads nicely to have all cases handled explicitly like this though? (I don't mind changing it if you prefer otherwise)